File: internal/agent/exec/utils/compare.go
package utils

import (
	"fmt"
	"github.com/aaronchen2k/deeptest/internal/pkg/consts"
	"strconv"
	"strings"
)

func Compare(operator consts.ComparisonOperator, actual, expect interface{}) (result consts.ResultStatus) {

	if operator == consts.Equal {
		if actual == expect {
			result = consts.Pass
		} else {
			result = consts.Fail
		}
	} else if operator == consts.NotEqual {
		if actual != expect {
			result = consts.Pass
		} else {
			result = consts.Fail
		}
	} else if operator == consts.Contain {
		if strings.Contains(interfaceToString(actual), interfaceToString(expect)) {
			result = consts.Pass
		} else {
			result = consts.Fail
		}
	} else {
		result = CompareAsFloat(operator, actual, expect)
	}

	return
}

func CompareAsFloat(operator consts.ComparisonOperator, actual, expect interface{}) (
	result consts.ResultStatus) {

	result = consts.Fail

	actualFloat, err1 := strconv.ParseFloat(interfaceToString(actual), 64)
	expectFloat, err2 := strconv.ParseFloat(interfaceToString(expect), 64)

	if err1 != nil || err2 != nil { // not a number
		return
	}

	switch operator.String() {
	case consts.GreaterThan.String():
		result = GetResult(actualFloat > expectFloat)

	case consts.LessThan.String():
		result = GetResult(actualFloat < expectFloat)

	case consts.GreaterThanOrEqual.String():
		result = GetResult(actualFloat >= expectFloat)

	case consts.LessThanOrEqual.String():
		result = GetResult(actualFloat <= expectFloat)

	default:

	}

	return
}

func GetResult(b bool) (
	result consts.ResultStatus) {

	if b {
		result = consts.Pass
	} else {
		result = consts.Fail
	}

	return

}

func interfaceToString(interf interface{}) (ret string) {
	ret = fmt.Sprintf("%v", interf)

	return
}

File: ui/src/utils/compare.ts
import {ComparisonOperator} from "./enum";

export const getCompareOpts = () => {
  const arr : any[] = []

  arr.push({label: ComparisonOperator.contain, value: ComparisonOperator[ComparisonOperator.contain]})
  arr.push({label: ComparisonOperator.notContain, value: ComparisonOperator[ComparisonOperator.notContain]})

  arr.push({label: ComparisonOperator.equal, value: ComparisonOperator[ComparisonOperator.equal]})
  arr.push({label: ComparisonOperator.notEqual, value: ComparisonOperator[ComparisonOperator.notEqual]})
  arr.push({label: ComparisonOperator.greaterThan, value: ComparisonOperator[ComparisonOperator.greaterThan]})
  arr.push({label: ComparisonOperator.greaterThanOrEqual, value: ComparisonOperator[ComparisonOperator.greaterThanOrEqual]})
  arr.push({label: ComparisonOperator.lessThan, value: ComparisonOperator[ComparisonOperator.lessThan]})
  arr.push({label: ComparisonOperator.lessThanOrEqual, value: ComparisonOperator[ComparisonOperator.lessThanOrEqual]})

  return arr
}

export const getCompareOptsForRespCode = () => {
  const arr : any[] = []

  arr.push({label: ComparisonOperator.equal, value: ComparisonOperator[ComparisonOperator.equal]})
  arr.push({label: ComparisonOperator.notEqual, value: ComparisonOperator[ComparisonOperator.notEqual]})

  return arr
}

export const getCompareOptsForString = () => {
  const arr : any[] = []

  arr.push({label: ComparisonOperator.contain, value: ComparisonOperator[ComparisonOperator.contain]})
  arr.push({label: ComparisonOperator.notContain, value: ComparisonOperator[ComparisonOperator.notContain]})
  arr.push({label: ComparisonOperator.equal, value: ComparisonOperator[ComparisonOperator.equal]})
  arr.push({label: ComparisonOperator.notEqual, value: ComparisonOperator[ComparisonOperator.notEqual]})

  return arr
}
